README
NAME
    File::Find::Upwards - Look for a file in the current directory and
    upwards

VERSION
    version 1.100860

SYNOPSIS
        use File::Find::Upwards qw(file_find_upwards);

        my $filename = file_find_upwards('myconfig.yaml');
        if ($filename) { rand() }

DESCRIPTION
    Provides functions that can find a file in the current or a parent
    directory.

FUNCTIONS
  file_find_upwards
    Takes a filename and looks for the file in the current directory. If
    there is no such file, it traverses up the directory hierarchy until it
    finds the file or until it reaches the topmost directory. If the file is
    found, the full path to the file is returned. If the file is not found,
    undef is returned.

    The result is memoized, so repeated calls to the function with the same
    filename will return the result of the first call for that filename.

    This function is exported automatically.

  find_containing_dir_upwards
    Like "file_find_upwards()", but reports the directory that contains the
    file. A "Path::Class::Dir" object is returned.

    This function is exported automatically.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=File-Find-Upwards>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/File-Find-Upwards/>.

    The development version lives at
    <http://github.com/hanekomu/File-Find-Upwards/>. Instead of sending
    patches, please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2008 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension File-Find-Upwards

1.100860  2010-03-27 11:15:47 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.04 Mon 2009.12.07 00:58:16 CET (Marcel Gruenauer <marcel@cpan.org>)
    - added find_containing_dir_upwards()
    - reduced dependencies using author tests
    - better dist engineering (INSTALL, README, README.mkdn etc.)

0.03  Tue Dec 16 13:20:03 CET 2008 (Marcel Gruenauer <marcel@cpan.org>)
    - Now based on Path::Class instead of File::Spec::Functions - trying to
      cleanly avoid a bug where it hanged on Windows because rootdir() there
      returned '\'

0.02 Sat Nov  1 09:17:35 CET 2008 (Marcel Gruenauer <marcel@cpan.org>)
    - Changed dist style and Changes back to standard. People didn't like it -
      the nail that sticks out gets hammered down.
    - Added standard test files; this will also help with CPANTS.

0.01  Thu, 29 May 2008 12:04:55 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - original version
    - .shipit: fixed svk.tagpattern and commit.header
    - Makefile.PL: added requirements
    - t/01_misc.t: added tests


lib/File/Find/Upwards.pm
use 5.008;
use strict;
use warnings;

package File::Find::Upwards;
our $VERSION = '1.100860';
# ABSTRACT: Look for a file in the current directory and upwards
use Path::Class;
use Attribute::Memoize;
use Exporter qw(import);
our @EXPORT = qw(file_find_upwards find_containing_dir_upwards);

sub file_find_upwards : Memoize {
    my $wanted_file = shift;
    my $dir         = dir('.')->absolute;
    my %seen;
    my $result;    # left undef as we'll return undef if we didn't find it
  LOOP: {
        do {
            last if $seen{$dir}++;
            my $file = $dir->file($wanted_file);
            if (-e $file) {
                $result = $file->absolute;
                last LOOP;
            }
        } while ($dir = $dir->parent);
    }
    $result;
}

sub find_containing_dir_upwards : Memoize {
    my $wanted_file = shift;
    my $dir         = dir('.')->absolute;
    my %seen;
    my $result;    # left undef as we'll return undef if we didn't find it
  LOOP: {
        do {
            last if $seen{$dir}++;
            my $file = $dir->file($wanted_file);
            if (-e $file) {
                return $dir;
            }
        } while ($dir = $dir->parent);
    }
    undef;
}
1;


__END__
=pod

=head1 NAME

File::Find::Upwards - Look for a file in the current directory and upwards

=head1 VERSION

version 1.100860

=head1 SYNOPSIS

    use File::Find::Upwards qw(file_find_upwards);

    my $filename = file_find_upwards('myconfig.yaml');
    if ($filename) { rand() }

=head1 DESCRIPTION

Provides functions that can find a file in the current or a parent directory.

=head1 FUNCTIONS

=head2 file_find_upwards

Takes a filename and looks for the file in the current directory. If there is
no such file, it traverses up the directory hierarchy until it finds the file
or until it reaches the topmost directory. If the file is found, the full path
to the file is returned. If the file is not found, undef is returned.

The result is memoized, so repeated calls to the function with the same
filename will return the result of the first call for that filename.

This function is exported automatically.

=head2 find_containing_dir_upwards

Like C<file_find_upwards()>, but reports the directory that contains the file.
A C<Path::Class::Dir> object is returned.

This function is exported automatically.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=File-Find-Upwards>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/File-Find-Upwards/>.

The development version lives at
L<http://github.com/hanekomu/File-Find-Upwards/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'File::Find::Upwards',
                       'DISTNAME' => 'File-Find-Upwards',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'Look for a file in the current directory and upwards',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100860',
                       'PREREQ_PM' => {
                                        'Scalar::Util' => '0',
                                        'English' => '0',
                                        'FindBin' => '0',
                                        'Exporter' => '0',
                                        'Attribute::Memoize' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'File::Find' => '0',
                                        'File::Temp' => '0',
                                        'Path::Class' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



